// @kinde-oss/kinde-typescript-sdk/dist/sdk/utilities/token-claims.js
/**
 * Method extracts the provided claim from the provided token type in the
 * current session.
 * @param {SessionManager} sessionManager
 * @param {string} claim
 * @param {ClaimTokenType} type
 * @returns {unknown | null}
 */
export const getClaimValue = async (sessionManager, claim, type = 'access_token') => {
    const tokenPayload = (await sessionManager.getSessionItem(`${type}_payload`));
    return tokenPayload[claim] ?? null;
};
/**
 * Method extracts the provided claim from the provided token type in the
 * current session, the returned object includes the provided claim.
 * @param {SessionManager} sessionManager
 * @param {string} claim
 * @param {ClaimTokenType} type
 * @returns {{ name: string, value: unknown | null }}
 */
export const getClaim = async (sessionManager, claim, type = 'access_token') => {
    return {
        name: claim,
        value: await getClaimValue(sessionManager, claim, type),
    };
};
/**
 * Method returns the organization code from the current session and returns
 * a boolean in the returned object indicating if the provided permission is
 * present in the session.
 * @param {SessionManager} sessionManager
 * @param {string} name
 * @returns {{ orgCode: string | null, isGranted: boolean }}
 */
export const getPermission = async (sessionManager, name) => {
    const permissions = ((await getClaimValue(sessionManager, 'permissions')) ??
        []);
    const isGranted = permissions.some((p) => p === name);
    const orgCode = (await getClaimValue(sessionManager, 'org_code'));
    return { orgCode, isGranted };
};
/**
 * Method extracts the organization code from the current session.
 * @param {SessionManager} sessionManager
 * @returns {{ orgCode: string | null }}
 */
export const getOrganization = async (sessionManager) => ({
    orgCode: (await getClaimValue(sessionManager, 'org_code')),
});
/**
 * Method extracts all the permission and the organization code in the access
 * token in the current session.
 * @param {SessionManager} sessionManager
 * @returns {{ permissions: string[], orgCode: string | null }}
 */
export const getPermissions = async (sessionManager) => {
    const [permissions, orgCode] = await Promise.all([
        (getClaimValue(sessionManager, 'permissions') ?? []),
        getClaimValue(sessionManager, 'org_code'),
    ]);
    return {
        permissions,
        orgCode,
    };
};
/**
 * Method extracts all organization codes from the id token in the current
 * session.
 * @param {SessionManager} sessionManager
 * @returns {{ orgCodes: string[] }}
 */
export const getUserOrganizations = async (sessionManager) => ({
    orgCodes: ((await getClaimValue(sessionManager, 'org_codes', 'id_token')) ??
        []),
});
