// @kinde-oss/kinde-typescript-sdk/dist/apis/BusinessApi.js
/* tslint:disable */
/* eslint-disable */
/**
 * Kinde Management API
 * Provides endpoints to manage your Kinde Businesses
 *
 * The version of the OpenAPI document: 1
 * Contact: support@kinde.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import * as runtime from '../runtime.js';
import { SuccessResponseFromJSON, } from '../models/index.js';
/**
 *
 */
export class BusinessApi extends runtime.BaseAPI {
    /**
     * Get your business details.
     * List business details
     */
    async getBusinessRaw(requestParameters, initOverrides) {
        if (requestParameters.code === null || requestParameters.code === undefined) {
            throw new runtime.RequiredError('code', 'Required parameter requestParameters.code was null or undefined when calling getBusiness.');
        }
        if (requestParameters.name === null || requestParameters.name === undefined) {
            throw new runtime.RequiredError('name', 'Required parameter requestParameters.name was null or undefined when calling getBusiness.');
        }
        if (requestParameters.email === null || requestParameters.email === undefined) {
            throw new runtime.RequiredError('email', 'Required parameter requestParameters.email was null or undefined when calling getBusiness.');
        }
        const queryParameters = {};
        if (requestParameters.code !== undefined) {
            queryParameters['code'] = requestParameters.code;
        }
        if (requestParameters.name !== undefined) {
            queryParameters['name'] = requestParameters.name;
        }
        if (requestParameters.email !== undefined) {
            queryParameters['email'] = requestParameters.email;
        }
        if (requestParameters.phone !== undefined) {
            queryParameters['phone'] = requestParameters.phone;
        }
        if (requestParameters.industry !== undefined) {
            queryParameters['industry'] = requestParameters.industry;
        }
        if (requestParameters.timezone !== undefined) {
            queryParameters['timezone'] = requestParameters.timezone;
        }
        if (requestParameters.privacyUrl !== undefined) {
            queryParameters['privacy_url'] = requestParameters.privacyUrl;
        }
        if (requestParameters.termsUrl !== undefined) {
            queryParameters['terms_url'] = requestParameters.termsUrl;
        }
        const headerParameters = {};
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("kindeBearerAuth", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/api/v1/business`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => SuccessResponseFromJSON(jsonValue));
    }
    /**
     * Get your business details.
     * List business details
     */
    async getBusiness(requestParameters, initOverrides) {
        const response = await this.getBusinessRaw(requestParameters, initOverrides);
        return await response.value();
    }
    /**
     * Update business details.
     * Update business details
     */
    async updateBusinessRaw(requestParameters, initOverrides) {
        if (requestParameters.businessName === null || requestParameters.businessName === undefined) {
            throw new runtime.RequiredError('businessName', 'Required parameter requestParameters.businessName was null or undefined when calling updateBusiness.');
        }
        if (requestParameters.primaryEmail === null || requestParameters.primaryEmail === undefined) {
            throw new runtime.RequiredError('primaryEmail', 'Required parameter requestParameters.primaryEmail was null or undefined when calling updateBusiness.');
        }
        const queryParameters = {};
        if (requestParameters.businessName !== undefined) {
            queryParameters['business_name'] = requestParameters.businessName;
        }
        if (requestParameters.primaryEmail !== undefined) {
            queryParameters['primary_email'] = requestParameters.primaryEmail;
        }
        if (requestParameters.primaryPhone !== undefined) {
            queryParameters['primary_phone'] = requestParameters.primaryPhone;
        }
        if (requestParameters.industryKey !== undefined) {
            queryParameters['industry_key'] = requestParameters.industryKey;
        }
        if (requestParameters.timezoneId !== undefined) {
            queryParameters['timezone_id'] = requestParameters.timezoneId;
        }
        if (requestParameters.privacyUrl !== undefined) {
            queryParameters['privacy_url'] = requestParameters.privacyUrl;
        }
        if (requestParameters.termsUrl !== undefined) {
            queryParameters['terms_url'] = requestParameters.termsUrl;
        }
        if (requestParameters.isShowKindeBranding !== undefined) {
            queryParameters['is_show_kinde_branding'] = requestParameters.isShowKindeBranding;
        }
        if (requestParameters.isClickWrap !== undefined) {
            queryParameters['is_click_wrap'] = requestParameters.isClickWrap;
        }
        if (requestParameters.partnerCode !== undefined) {
            queryParameters['partner_code'] = requestParameters.partnerCode;
        }
        const headerParameters = {};
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("kindeBearerAuth", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/api/v1/business`,
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => SuccessResponseFromJSON(jsonValue));
    }
    /**
     * Update business details.
     * Update business details
     */
    async updateBusiness(requestParameters, initOverrides) {
        const response = await this.updateBusinessRaw(requestParameters, initOverrides);
        return await response.value();
    }
}
