// @kinde-oss/kinde-typescript-sdk/dist-cjs/apis/ConnectedAppsApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Kinde Management API
 * Provides endpoints to manage your Kinde Businesses
 *
 * The version of the OpenAPI document: 1
 * Contact: support@kinde.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectedAppsApi = void 0;
const runtime = __importStar(require("../runtime.js"));
const index_js_1 = require("../models/index.js");
/**
 *
 */
class ConnectedAppsApi extends runtime.BaseAPI {
    /**
     * Get a URL that authenticates and authorizes a user to a third-party connected app.
     * Get Connected App URL
     */
    async getConnectedAppAuthUrlRaw(requestParameters, initOverrides) {
        if (requestParameters.keyCodeRef === null || requestParameters.keyCodeRef === undefined) {
            throw new runtime.RequiredError('keyCodeRef', 'Required parameter requestParameters.keyCodeRef was null or undefined when calling getConnectedAppAuthUrl.');
        }
        if (requestParameters.userId === null || requestParameters.userId === undefined) {
            throw new runtime.RequiredError('userId', 'Required parameter requestParameters.userId was null or undefined when calling getConnectedAppAuthUrl.');
        }
        const queryParameters = {};
        if (requestParameters.keyCodeRef !== undefined) {
            queryParameters['key_code_ref'] = requestParameters.keyCodeRef;
        }
        if (requestParameters.userId !== undefined) {
            queryParameters['user_id'] = requestParameters.userId;
        }
        const headerParameters = {};
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("kindeBearerAuth", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/api/v1/connected_apps/auth_url`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, index_js_1.ConnectedAppsAuthUrlFromJSON)(jsonValue));
    }
    /**
     * Get a URL that authenticates and authorizes a user to a third-party connected app.
     * Get Connected App URL
     */
    async getConnectedAppAuthUrl(requestParameters, initOverrides) {
        const response = await this.getConnectedAppAuthUrlRaw(requestParameters, initOverrides);
        return await response.value();
    }
    /**
     * Get an access token that can be used to call the third-party provider linked to the connected app.
     * Get Connected App Token
     */
    async getConnectedAppTokenRaw(requestParameters, initOverrides) {
        if (requestParameters.sessionId === null || requestParameters.sessionId === undefined) {
            throw new runtime.RequiredError('sessionId', 'Required parameter requestParameters.sessionId was null or undefined when calling getConnectedAppToken.');
        }
        const queryParameters = {};
        if (requestParameters.sessionId !== undefined) {
            queryParameters['session_id'] = requestParameters.sessionId;
        }
        const headerParameters = {};
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("kindeBearerAuth", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/api/v1/connected_apps/token`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, index_js_1.ConnectedAppsAccessTokenFromJSON)(jsonValue));
    }
    /**
     * Get an access token that can be used to call the third-party provider linked to the connected app.
     * Get Connected App Token
     */
    async getConnectedAppToken(requestParameters, initOverrides) {
        const response = await this.getConnectedAppTokenRaw(requestParameters, initOverrides);
        return await response.value();
    }
    /**
     * Revoke the tokens linked to the connected app session.
     * Revoke Connected App Token
     */
    async revokeConnectedAppTokenRaw(requestParameters, initOverrides) {
        if (requestParameters.sessionId === null || requestParameters.sessionId === undefined) {
            throw new runtime.RequiredError('sessionId', 'Required parameter requestParameters.sessionId was null or undefined when calling revokeConnectedAppToken.');
        }
        const queryParameters = {};
        if (requestParameters.sessionId !== undefined) {
            queryParameters['session_id'] = requestParameters.sessionId;
        }
        const headerParameters = {};
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("kindeBearerAuth", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/api/v1/connected_apps/revoke`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, index_js_1.SuccessResponseFromJSON)(jsonValue));
    }
    /**
     * Revoke the tokens linked to the connected app session.
     * Revoke Connected App Token
     */
    async revokeConnectedAppToken(requestParameters, initOverrides) {
        const response = await this.revokeConnectedAppTokenRaw(requestParameters, initOverrides);
        return await response.value();
    }
}
exports.ConnectedAppsApi = ConnectedAppsApi;
